Implementation of the ARIA cipher

pauli-oracle opened the pull request:
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

##### Description of change

Implementation of the ARIA cipher as described in RFC 5794.  This
implementation is written in endian agnostic C code.  No attempt
at providing machine specific assembly code has been made.  This
implementation expands the _evptests_ by including the test cases from
RFC 5794 rather than providing an individual test case.  Support for
ARIA has been integrated into the command line applications.

Some of the ARIA cipher suites defined in RFC 6209 have been implemented.
This does not include suites which would be considered too weak or the
GCM based suites.  ARIA based TLS has been validated against CycloneTLS
and cross validated between Solaris/SPARC, Linux/390 and Linux/x64.

Fixes #2336.

richsalz commented:
This has an Oracle copyright in it.  Therefore we will need a CCLA from Oracle, and an ICLA from all people who wrote the code.

pauli-oracle commented:
Oracle should have a CCLA already.  At least our legal department has told us we do.
I submitted my ICLA along with this.  I'm the only author.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+    unsigned int i;
+    ARIA_u128 t;
+
+    for (i=0; i<ARIA_BLOCK_SIZE; i++)
```
same thing about spaces; will not comment on it further.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+    ARIA_u128 p;
+    unsigned int i;
+
+    memcpy(p, pin, sizeof(ARIA_u128));
```
```sizeof(p)``` for 3rd arg.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+ * Diffusion layer step
+ * It is NOT safe for the input and output to overlap.
+ */
+static void a(ARIA_u128 y, const ARIA_u128 x) {
```
curly brace on its own lne.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+        t[(i+bytes) % ARIA_BLOCK_SIZE] = z[i];
+    for (i=0; i<ARIA_BLOCK_SIZE; i++)
+        o[i] = ((t[i] >> bits) |
+               (t[i?i-1:ARIA_BLOCK_SIZE-1] << (8 - bits))) ^ xor[i];
```
space around operators.  I won't comment on it further.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+void ARIA_encrypt(const unsigned char *in, unsigned char *out,
+                        const ARIA_KEY *key)
+{
+    assert(in && out && key);
```
Explicit ```!=  NULL``` tests.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+        return -1;
+    memcpy(w0, userKey, sizeof(ARIA_u128));
+    switch(bits) {
+    default: return -2;
```
return on its own line.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+
+    if (!userKey || !key)
+        return -1;
+    memcpy(w0, userKey, sizeof(ARIA_u128));
```
we use sizeof the destination, so ```sizeof(w0)``` here.  will not comment further.

richsalz commented on crypto/aria/aria.c:
```diff
@@ -0,0 +1,462 @@
+
+        if (key->rounds > 14) {
+            rot31l(key->rd_key[15], w3, w0);
+
```
delete this blank line.

richsalz commented on include/openssl/aria.h:
```diff
@@ -0,0 +1,68 @@
+#ifndef HEADER_ARIA_H
```
copyright missing.

pauli-oracle replied:
This batch should be finished.  Thanks for the comments.

richsalz commented:
This cipher should be disabled by default.
I am not sure we want to be adding more CBC ciphers these days, only AEAD constructs unless there is a (large) body of Aria-encrypted data at rest.

dot-asm commented on util/libcrypto.num:
```diff
@@ -4231,3 +4231,31 @@ X509_VERIFY_PARAM_get_time              4181	1_1_0d	EXIST::FUNCTION:
 EVP_PKEY_get0_poly1305                  4182	1_1_1	EXIST::FUNCTION:POLY1305
 DH_check_params                         4183	1_1_0d	EXIST::FUNCTION:DH
 EVP_PKEY_get0_siphash                   4184	1_1_1	EXIST::FUNCTION:SIPHASH
+ARIA_set_decrypt_key                    4185	1_1_1	EXIST::FUNCTION:ARIA
```
No, no and no. No `ARIA_*` symbols should be exported from libcrypto shared library. Cipher should be accessible only through EVP. Just like every new algorithm. This does mean that aria.h should not be in include/openssl, that aria_{cbc|cfb|ctr|ecb|ofb}.c are essentially redundant, because corresponding functionality can be implemented directly in e_aria.c. [And it also means that there is no need for apps/speed.c modifications (speed -evp should be used)].

pauli-oracle commented:
Thanks for the feedback Andy, I've a better understanding of what is required now.  I'll make the changes you've requested.

richsalz commented:
We have your ICLA and the Oracle CCLA, so that's all set.

pauli-oracle commented:
Thanks for sorting out the CCLA.

I've disabled ARIA by default, the _enable-aria_ configuration option is now required to turn it on.

I understand the concern about using ARIA-CBC.  I know that it is used by the Oracle Database which means there is likely a very large body of data involved.  I will try to find out more.

There is also a case for using CBC in the context of TLS. All of the cipher suites defined by RFC 6209 use either CBC or GCM for chaining.

dot-asm commented:
> it is used by the Oracle Database which means there is likely a very large body of data involved

This is quite weak argument. Question is how would OpenSSL implementation help those users who have data at rest? Does Oracle document how exactly is the data in database container encrypted [and keys managed]? Or in other words do users actually have a choice to not use Oracle Database software to access their data? Because if they don't, then argument is essentially irrelevant. Unless of course Oracle plans to switch to OpenSSL in Database...

As for TLS and suggested cipher suites. I agree with Rich and would also say that GCM should be the only one available through TLS. Though even then question about relevancy remains. Are there [or planned] TLS servers that don't offer anything else but ARIA? If planned is this submission part of that plan?

pauli-oracle commented:
Oracle does plan to switch to OpenSSL in Database. I have received confirmation that CBC is the only mode used by South Korean customers for both data at rest and network encryption.

dot-asm commented:
> CBC is the only mode used by South Korean customers for both data at rest and network encryption.

Do note that I didn't say "discard CBC implementation" but that "GCM should be the only one available through TLS." I.e. it's not about general availability of CBC mode to *an* application, but about its exposure specifically through TLS. As for CBC being used for network encryption, assuming through TLS and not some custom protocol. Yes, one can naturally argue in favour of providing CBC for backward compatibility with current clients. But at the same time it's as important to define a vector for desired future development. And consensus (as effectively reflected in upcoming TLS specification) is that CBC is has no place in future. Also note that question was whether or not there are TLS peers that negotiate *exclusively* ARIA-CBC. Background is that maybe it doesn't *have to* be implemented (again, at TLS level), as they might be free to negotiate something else. Well, it's probably about some government regulations, i.e. some peers might be *obliged* to negotiate specific cipher. Is this it? [BTW, what do they use right now?] But either way this is not argument *against* implementing GCM, only in favour of providing CBC at TLS level.

dot-asm commented on crypto/include/internal/aria.h:
```diff
@@ -0,0 +1,56 @@
+
+
+extern int aria_set_encrypt_key(const unsigned char *userKey, const int bits,
+                        ARIA_KEY *key);
```
I don't follow why is it declared `extern`. It can only be styling question, but it's not our style. But our styling rules say that second line above should be aligned. And I suppose if you drop `extern` it will be aligned...

pauli-oracle replied:
I've dropped the externs -- just a habit of mine more than anything and aligned the second lines.

dot-asm commented on test/evptests.txt:
```diff
@@ -1604,6 +1604,45 @@ Operation = ENCRYPT
+Key = 000102030405060708090a0b0c0d0e0f
+Operation = DECRYPT
+Plaintext = 00112233445566778899aabbccddeeff
+Ciphertext = d718fbd6ab644c739da95f3be6451778
```
If you omit `Operation` then it tests both ways automatically. In other words you need only one copy of each vector. It's also easier to double-check against RFC text.

pauli-oracle replied:
Thanks for this, I wasn't aware.

dot-asm commented:
I thought that speed.c modifications was reverted, otherwise why do I still see it in "Files changed"?

pauli-oracle commented:
Thanks for the detailed comment, I agree that GCM is a better choice than CBC in general.  I implemented ARIA-CBC partly for completeness and partly because I thought it was currently being used.  However, it has just become evident that we do not use it at Oracle.  I've updated this request with the relevant code removed.

I would be happy to implement GCM cipher suites as a distinct PR later.

dot-asm commented:
Side note. Be nice to CI, add [skip ci] to commit text for formatting updates :-) Well, I suppose this is for future. I'll have to have second look at whole thing...

pauli-oracle commented:
That for the CI hint.  This is definitely a learning experience :)

dot-asm commented:
Squash all commits except for "spelling of Camellia" [as it's not actually related] to one, re-base to master to resolve conflict, and force-push it. BTW, those files that ended up with just year updates, it's probably appropriate to omit them to reduce noise. As for test vectors. ARIA home page has more vectors for different modes, it would be appropriate to add them too...

pauli-oracle commented:
Everything is squashed down now.

The extra test vectors have been added to evptests.

I've removed the copyright year changes on the perl script that wasn't producing different output anymore.  I left them there on the couple of scripts that did produce different output.

dot-asm commented:
I've double-checked test vectors against published one and 7c84a1c0a4932d4f83b6ef179af919649915168eff8c49b6010c9fdf17378692 is the sha256 hash for test/evptests.txt. Red cross in appveyor is unrelated. enable-aria --strict-warnings build passes.

dot-asm commented on ssl/ssl_ciph.c:
```diff
@@ -101,10 +101,7 @@ static const ssl_cipher_table ssl_cipher_table_cipher[SSL_ENC_NUM_IDX] = {
-    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL,
-    NULL, NULL
-};
+static const EVP_CIPHER *ssl_cipher_methods[SSL_ENC_NUM_IDX];
```
Speaking of irony. I'm not 100% sure why it's explicitly initialized with NULLs, but I do recall problem with Solaris long time ago, which was attributed to failure to initialize static variables to zero under some circumstances. What I'm saying is that this modification should be omitted too, but not necessarily because Sun C failed to comply with standard long time ago, but because it's unrelated modification.

On related note. As curious fact, i.e. not related to the above, because ssl_cipher_methods itself is not constant. But `const int foo[4];` and `const int foo[4] = {};` are compiled differently and end up either in .bss or .rodata segments.

pauli-oracle replied:
I made this change based on a comment by Rich back in January :)  I've removed it now.  I didn't realise there were repercussions on some platforms -- BTW the initialisation list is too short for the array so currently there is a mixture of explicit and implicit initialisers.

I'd not noticed the different segments for the two declarations.  That's interesting.  A little weird but gcc often can be.

dot-asm replied:
This is not specific to gcc, for example Sun C does exactly same. In other words if one cares about data placement you have to watch for initializers... But once again, suggestion to remove was based on the fact that it was unrelated.

pauli-oracle commented:
That SHA256 hash matches the file I've got here.

dot-asm commented on doc/man1/ciphers.pod:
```diff
@@ -579,6 +584,21 @@ Note: these ciphers can also be used in SSL v3.
+ TLS_ECDHE_ECDSA_WITH_ARIA_256_CBC_SHA384  ECDHE-ECDSA-ARIA256-CBC-SHA384
+ TLS_ECDHE_RSA_WITH_ARIA_128_CBC_SHA256    ECDHE-RSA-ARIA128-CBC-SHA256
+ TLS_ECDHE_RSA_WITH_ARIA_256_CBC_SHA384    ECDHE-RSA-ARIA256-CBC-SHA384
+
```
Oh! I suppose ciphers.pod modification should be omitted too. At least till GCM is implemented...

dot-asm commented:
Modulo ciphers.pod I'd say it's ready to go. On related note, it would probably be appropriate to add enable-aria to one of CI targets...

pauli-oracle commented:
Fixed the pod issue.
Added ARIA to one of the CI loops, I wasn't clear which was most appropriate so I didn't squash this one done.

pauli-oracle commented:
Added ARIA to one of the CI builds: https://travis-ci.org/openssl/openssl/jobs/200983711

dot-asm commented:
+1 with reservation that we can remodel the commit message. Essentially remove squashing artefacts, and maybe add couple of words on modes availability...

pauli-oracle commented:
I'm not clear what is being asked for.  Do you mean just the first paragraph and a description of the modes implemented?  Or do you want the first paragraph trimmed back to something like "implementation of the ARIA cipher as per RFC 5794 with these cipher modes ..."

dot-asm commented:
You are asked to not be surprised (or complain :-) when you see that commit *message* to repository is different from one available in this merge request. In other words we reserve the right to modify it, withing reasonable limits.

pauli-oracle commented:
I understand now, you don't want me to modify it, which is great :)

To save searching, the following cipher modes are supported with the three key sizes of ARIA: CBC, CFB1, CFB8, CFB128, CTR, ECB & OFB128.

richsalz commented:
@dot-asm you going to merge or me?

dot-asm commented:
Merged. Thanks for contribution and patience.
